web/src/app/components/cdf/cdf.utils.ts

```diff
@@ -0,0 +1,294 @@
+ */
+function getDurationforProbability(data: Item[], probability: number): number {
+  const allProbabilities = data.map(d => d.probability);
+  const left = data[d3.bisectLeft(allProbabilities, probability) - 1];
+  const right = data[d3.bisectRight(allProbabilities, probability)];
+  return ((probability - left.probability) * (right.duration - left.duration) /
+    (right.probability - left.probability)) + left.duration;
+}
+
+/**
+ * Generates an array of items that maps an array of probabilities to its
+ * interpolated duration. If the durations are too close together, then
+ * recursively spread out the smallest and largest probability line until there
+ * is at least 15 pixels of space between them. If this never happens, then the
+ * smallest and largest probabilities are deleted and only the median is
+ * returned.
+ *
+ * @param data Array of Items sorted by increasing duration
+ * @param percentileLines Array of numbers representing the probability
+ * @param xScale d3 function that applies a scaling factor on raw x values to
+ * correctly place them on the graph
+ * @return Array of Items with the duration as the interpolated duration and the
+ * probability as the given probability from probabilityLines
+ */
+export function generateQuantiles(
+                  data: Item[],
+                  percentileLines: number[],
+                  xScale: d3.ScaleLinear<number, number>): Item[] {
+  if (percentileLines[0] < 0.01 || percentileLines[2] > .99) {
+    return [percentileLines[1]].map(d => ({
+      duration: getDurationforProbability(data, d),
+      probability: d} as Item));
+  }
+  let quantiles = percentileLines.map(d => ({
+    duration: getDurationforProbability(data, d),
+    probability: d} as Item));
+
+  const pixelDifference = 15;
+  const differences = [xScale(quantiles[1].duration - quantiles[0].duration),
+                      xScale(quantiles[2].duration - quantiles[1].duration)];
+  if (differences[0] < pixelDifference || differences[1] < pixelDifference) {
+    quantiles = generateQuantiles(
+                  data,
+                  [percentileLines[0] - .01,
+                  percentileLines[1],
+                  percentileLines[2] + .01],
+                  xScale);
+  }
+  return quantiles;
+}
+
+/**
+ * Produces an array of numbers that represent the amount of height needed to
+ * place a dot at that location. Looping through the duration values in sorted
+ * order, if placing a dot overlaps with previously placed dots, then increase
+ * the y position to dodge that dot. This loops until the (x, y) position of the
+ * does not overlap with any other dots.
+ *
+ * @param radius radius of dots
+ * @param xScale d3 function that applies a scaling factor on raw x values to
+ * correctly place them on the graph
+ * @param xVals array of durations in sorted order
+ * @return Array of numbers representing the y position of the dot at duration
+ * value maintaining the same indices as the xVals
```

Comment:
```suggestion
 * @param radius radius of dots
 * @param xScale d3 function that applies a scaling factor on raw x values to
 *     correctly place them on the graph
 * @param xVals array of durations in sorted order
 * @return Array of numbers representing the y position of the dot at duration
 *     value maintaining the same indices as the xVals
```

The continuation should use and indentation of 4 spaces. Same comment for the other similar places.

Reference: https://google.github.io/styleguide/jsguide.html#jsdoc-line-wrapping